src/braket/reservations/reservations.py

```diff
@@ -0,0 +1,72 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from contextlib import contextmanager
+
+from braket.aws.aws_device import AwsDevice, AwsDeviceType
+from braket.devices.device import Device
+from braket.devices.local_simulator import LocalSimulator
+
+
+@contextmanager
+def Reservation(device: Device, reservation_arn: str | None = None) -> Device | None:
+    """
+    A context manager that temporarily modifies the `run` method of a `device` object
+    to include a specified `reservation_arn` in all calls within the context. This is useful
+    for ensuring that all quantum task run within the context are associated with a given
+    reservation ARN (Amazon Resource Name).
+
+    Args:
+        device (Device): The Braket device for which you have a reservation ARN.
+        reservation_arn (str | None): The Braket Direct reservation ARN to be applied to all
+            quantum tasks run within the context. Defaults to None.
+
+    Yields:
+        Device | None: The device object with a potentially modified `run` method. If the device is
+        a local simulator or the reservation ARN is None, the original device is yielded without
+        modification.
+
+    Examples:
+        >>> with Reservation(device, reservation_arn="<my_reservation_arn>"):
```

Comment:
Minor nit: can you also add an example of this used as a decorator?

Reply:
interesting, I did not know the context could also be a decorator. Why don't we also have Tracker available as a decorator?

Reply:
I would have to ask around for why that was not the case. Could be that the functionality in `contextmanager` was not known at the time.